pkg/parser/io_parser.go

```diff
@@ -0,0 +1,152 @@
+package parser
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"sync"
+)
+
+const (
+	newline    = '\n'
+	newlineLen = 1
+)
+
+type IOParser struct {
+	inputBuffer  []byte
+	inputLock    sync.Mutex
+	outputBuffer []byte
+	outputLock   sync.Mutex
+	parse        *Parser
+}
+
+func NewIOParser(bufSize uint64, parserBufSize uint64, debug bool) *IOParser {
+	return &IOParser{
+		inputBuffer:  make([]byte, 0, bufSize),
+		outputBuffer: make([]byte, 0, bufSize),
+		parse:        NewBufferedParser(parserBufSize, debug),
+	}
+}
+
+func (p *IOParser) Read(b []byte) (int, error) {
+	p.outputLock.Lock()
+	defer p.outputLock.Unlock()
+
+	var (
+		bufcap    = cap(b)
+		outbuflen = len(p.outputBuffer)
+	)
+
+	// If somehow we have exactly cap(b) in output buffer, just copy and return.
+	if outbuflen == bufcap {
+		copy(b, p.outputBuffer)
```

Comment:
Isn't it wrong to use `cap(b)` when calculating these copies? The [io.Reader](https://golang.org/pkg/io/#Reader) API specifies that `Read reads up to len(p) bytes into p.` Where `p` is our `b`. Same also applies to the builtin `copy`.

So if I understand correctly, if the `bufcap` matched `outbuflen` but `len(b) == 0` then nothing is copied, but the code thinks something has been copied.

If I understand the copy section before the for-loop below. This is going to fill `b` with buffered data as much as it can. And when the buffer runs out it reads more. So could the implementation be more akin to:

```
written := copy(b, p.outputBuffer)
p.outputBuffer = p.outputBuffer[n:] // maybe copying the buffer to the start of the underlying slice buffer as at L52
if written == len(b) {
  return n, nil
}
```

Same buf vs len discussion applies to all code below.

Reply:
> Isn't it wrong to use cap(b) when calculating these copies? The io.Reader API specifies that Read reads up to len(p) bytes into p. Where p is our b. Same also applies to the builtin copy.

Hey, that's a good point, and a simple fix. In reality, 99% of the cases are such that `cap(b) = len(b)`, but it is not a guarantee, and as you said, the interface explicitly requires relying on length, rather than capacity 👍 

> If I understand the copy section before the for-loop below ...

Yes, this should work!

Reply:
Updated.